### PR TITLE
JAMES-2321 WebAdmin should allow to see more details about Mails

### PR DESCRIPTION
**Pull request by blackheaven**
Continuation of #1454 (for CI)

### CODE REVIEW

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 1]
```diff
@@ -19,39 +19,208 @@
...
+            fetchBody(additionalFields, mail),
+            fetchMessageSize(additionalFields, mail));
+    }
+
+    private static Optional<Long> fetchMessageSize(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.MESSAGE_SIZE)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.of(new Long(mail.getMessageSize()));
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Is that wrapping needed?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 2]
```diff
@@ -19,39 +19,208 @@
...
+    }
+
+    private static Optional<HeadersDto> fetchHeaders(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.HEADERS)) {
+            return Optional.empty();
+        }
+
+        try {
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function((MimeMessage message) -> {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Big lambda's like this are hardly readable and should be extracted...

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 3]
```diff
@@ -19,39 +19,208 @@
...
+                        HeadersDto headers = new HeadersDto();
+                        Collections
+                            .list(message.getAllHeaders())
+                            .forEach((header) -> headers.add(header.getName(), header.getValue()));
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Still some mutability around? Please do use `Steam::collect` with the `Guavate.toImmutableListMultimap()`, you gonna love it!

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/HeadersDto.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import com.google.common.collect.Multimap;
+
+public class HeadersDto {
+    private Multimap<String, String> headers;
+
+    public HeadersDto() {
+        headers = ArrayListMultimap.create();
+    }
+
+    public void add(String key, String value) {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/HeadersDto.java]: DTO should not be mutable... take that multimap with all the entries directly in the constructor.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 5]
```diff
@@ -19,39 +19,208 @@
...
+
+    private static Optional<Map<String, HeadersDto>> fetchPerRecipientsHeaders(List<AdditionalField> additionalFields, Mail mail) {
+        if (!additionalFields.contains(AdditionalField.PER_RECIPIENTS_HEADERS)) {
+            return Optional.empty();
+        }
+        Map<String, HeadersDto> headers = new HashMap<>();
+        PerRecipientHeaders specificHeaders = mail.getPerRecipientSpecificHeaders();
+        Multimap<MailAddress, PerRecipientHeaders.Header> headersByRecipient = specificHeaders.getHeadersByRecipient();
+
+        for (MailAddress address : headersByRecipient.keySet()) {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Let's deprecate for. Use stream, collect, filter, map, flatMap etc.. And avoid side effects.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 6]
```diff
@@ -19,39 +19,208 @@
...
+        }
+
+        return Optional.of(headers);
+    }
+
+    private static HeadersDto fetchPerRecipientHeader(
+            Multimap<MailAddress, PerRecipientHeaders.Header> headersByRecipient,
+            MailAddress address) {
+        HeadersDto header = new HeadersDto();
+        headersByRecipient.get(address).forEach((rawHeader) -> header.add(rawHeader.getName(), rawHeader.getValue()));
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Idem

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 7]
```diff
@@ -19,39 +19,208 @@
...
+        return header;
+    }
+
+    private static Optional<Map<String, String>> fetchAttributes(List<AdditionalField> additionalFields, Mail mail) {
+        if (!additionalFields.contains(AdditionalField.ATTRIBUTES)) {
+            return Optional.empty();
+        }
+
+        return Optional.of(Iterators.toStream(mail.getAttributeNames())
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Ident should be space

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 8]
```diff
@@ -399,4 +418,20 @@ private void enforceActionParameter(Request request) {
     private String decodedRepositoryUrl(Request request) throws UnsupportedEncodingException {
         return URLDecoder.decode(request.params("encodedUrl"), StandardCharsets.UTF_8.displayName());
     }
+
+    private List<AdditionalField> extractAdditionalFields(String additionalFieldsParam) throws IllegalArgumentException {
+        List<AdditionalField> additionalFields = new ArrayList<>();
+
+        for (String requestedField : Splitter.on(',').trimResults().omitEmptyStrings().split(additionalFieldsParam)) {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: Use streams, and no for loop + avoid side effects.

Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: The splitter line can be extracted, named and indented, right?

Also you can use `.splitToList()` it gonna save you some pain.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 9]
```diff
@@ -399,4 +418,20 @@ private void enforceActionParameter(Request request) {
...
+            Optional<AdditionalField> correspondingField = AdditionalField.find(requestedField);
+            if (correspondingField.isPresent()) {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: a isPresent in a if is generally a design smell. You should use optional as a monad...

For the record here:

```
additionalFields.add(correspondingField
    .orElseThrow(() -> new IllegalArgumentException(requestedField)));
```

Does the same thing than your code

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 10]
```diff
@@ -427,9 +445,159 @@ public void retrievingAMailShouldDisplayItsInformation() throws Exception {
...
+            .extract()
+                .body()
+                .asString();
+
+        assertThatJson(jsonAsString)
+            .when(IGNORING_ARRAY_ORDER)
+            .when(IGNORING_EXTRA_FIELDS)
+            .ignoring("body")
+                .isEqualTo("{\n" +
+                        "  \"name\": \"name1\",\n" +
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: Still these `\n` ?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java (hunk 7 above)
Comment (blackheaven) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: It's already space-indented, so I'm a bit lost

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 11]
```diff
@@ -19,39 +19,208 @@
...
             Optional.ofNullable(mail.getErrorMessage()),
-            Optional.ofNullable(mail.getState()));
+            Optional.ofNullable(mail.getState()),
+            Optional.ofNullable(mail.getRemoteHost()),
+            Optional.ofNullable(mail.getRemoteAddr()),
+            Optional.ofNullable(mail.getLastUpdated()),
+            fetchAttributes(additionalFields, mail),
+            fetchPerRecipientsHeaders(additionalFields, mail),
+            fetchHeaders(additionalFields, mail),
+            fetchBody(additionalFields, mail),
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Yes, that is!

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java (hunk 7 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Should be 4 space, not more.

Sorry I was numlock and did not saw it.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (blackheaven) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Just to be sure: you want that the output message contains either `bodyText` or `bodyHtml`, contrarly to the the issue which only expect `body`?

[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: I want it to contain both, them being potentially empty.

Like : 

```
{
...
     "textBody": "A text version of my message",
     "htmlBody": null // this message has no HTML
...
}

{
...
     "textBody": null,   // this message has no plain text version provided
     "htmlBody": "<body>awesome HTML only message</body>"
...
}
 
{  // this message has both!
...
     "textBody": "A text version of my message",
     "htmlBody": "<body>awesome HTML only message</body>"
...
}

{ // this message has none
...
     "textBody": null,
     "htmlBody": null 
...
}
```

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 12]
```diff
@@ -19,39 +19,205 @@
...
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
+                    .map(Throwing.function(extractor::extract).sneakyThrow())
+                    .flatMap(MailDto::extractBody);
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.BODY, e);
+        }
+    }
+
+    private static Optional<String> extractBody(MessageContent content) {
+        Optional<String> body = content.getTextBody();
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: could be ```OptionalUtils.orSuppliers(content::getTextBody, content::getHtmlBody)``` ?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 13]
```diff
@@ -19,39 +19,205 @@
...
+
+    private static Optional<String> fetchBody(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.BODY)) {
+            return Optional.empty();
+        }
+
+        try {
+            MessageContentExtractor extractor = new MessageContentExtractor();
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: we build a Dto to not transfer it ? Didn't we abuse it because it has a handy feature we want ?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 14]
```diff
@@ -19,39 +19,205 @@
...
+            Optional.ofNullable(mail.getRemoteAddr()),
+            Optional.ofNullable(mail.getLastUpdated()),
+            fetchAttributes(additionalFields, mail),
+            fetchPerRecipientsHeaders(additionalFields, mail),
+            fetchHeaders(additionalFields, mail),
+            fetchBody(additionalFields, mail),
+            fetchMessageSize(additionalFields, mail));
+    }
+
+    private static Optional<Long> fetchMessageSize(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: additionalFields could be a Set because we keep searching into it and we don't want duplicate entries anyway

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 15]
```diff
@@ -406,19 +417,26 @@ public void retrievingRepositorySizeShouldReturnNumberOfContainedMails() throws
     @Test
     public void retrievingAMailShouldDisplayItsInformation() throws Exception {
         when(mailRepositoryStore.get(URL_MY_REPO)).thenReturn(Optional.of(mailRepository));
-
+        SimpleDateFormat dateTimeFormatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: when it comes to Dates, you should rather use java.time package.
In this case, It looks like you want ISO_INSTANT from `DateTimeFormatter`

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 16]
```diff
@@ -406,19 +417,26 @@ public void retrievingRepositorySizeShouldReturnNumberOfContainedMails() throws
...
+        dateTimeFormatter.setTimeZone(TimeZone.getTimeZone("UTC"));
         String name = NAME_1;
         String sender = "sender@domain";
         String recipient1 = "recipient1@domain";
         String recipient2 = "recipient2@domain";
         String state = "state";
         String errorMessage = "Error: why this mail is stored";
+        String remoteHost = "smtp.domain";
+        String remoteAddr = "66.66.66.66";
+        Date lastUpdated = new Date();
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: we try  to write repeatable tests, so using a arbitrary date is better in this respect

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 17]
```diff
@@ -427,9 +445,159 @@ public void retrievingAMailShouldDisplayItsInformation() throws Exception {
...
+            .body("sender", is(sender))
+            .body("headers", nullValue())
+            .body("body", nullValue())
+            .body("messageSize", is(messageSize))
+            .body("attributes", nullValue())
+            .body("perRecipientsHeaders", nullValue());
+    }
+
+    @Test
+    public void retrievingAMailShouldFailWhenAnUnknownFieldIsRequested() throws Exception {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: you should check duplicate fields in another test

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java (hunk 13 above)
Comment (blackheaven) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: I'm not sure to get it, can you re-explain it please?

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java (hunk 15 above)
Comment (blackheaven) [reply server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: The TimeZone is missing but I take the idea

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/InnaccessibleFieldException.java [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+import javax.mail.MessagingException;
+
+import org.apache.james.webadmin.dto.MailDto.AdditionalField;
+
+public class InnaccessibleFieldException extends Exception {
+
+    private final AdditionalField field;
+    private final MessagingException originalException;
+
+    public InnaccessibleFieldException(AdditionalField field, MessagingException originalException) {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/InnaccessibleFieldException.java]: `originalException` sounds like `cause` to me.

It should be passed to the parent constructor to be part of the stacktrace, and generic.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java (hunk 13 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Yes @mbaechler  there is a mis-understanding here: `convertMessage` does convert the message to MIME4J message. It do not build a DTO.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 19]
```diff
@@ -19,39 +19,217 @@
...
+        }
+    }
+
+    private static Optional<String> fetchTextBody(Set<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.TEXT_BODY)) {
+            return Optional.empty();
+        }
+
+        try {
+            MessageContentExtractor extractor = new MessageContentExtractor();
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: That, is symptomatic from the fact that we need a `MailDTOFactory` class: We have way too much static methods, which creates objects on every call objects that could be re-used.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 20]
```diff
@@ -19,39 +19,217 @@
...
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
+                    .map(Throwing.function(extractor::extract).sneakyThrow())
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: It's a shame... MIME parsing is memory and time expensive, we should ensure we do parse that message only once, in `fromMail` then only use the MessageCOntent in `fetchHtmlBody` & `fetchTextBody`.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 21]
```diff
@@ -19,39 +19,217 @@
...
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::extractHeaders).sneakyThrow());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.HEADERS, e);
+        }
+    }
+
+    private static HeadersDto extractHeaders(MimeMessage message) throws MessagingException {
+        return new HeadersDto(Collections
+            .list(message.getAllHeaders())
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: What about header unscrambling in HeaderDTO? 

`MimeUtil.unscrambleHeaderValue(headerValue)`

Overwise we will encounter encoding issues...

Comment (blackheaven) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: Headers can have encoding?

Do you mean I have to extract this method in MimeUtils?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 22]
```diff
@@ -51,6 +51,7 @@
 
 public class MailDto {
     public static MailDto fromMail(Mail mail, Set<AdditionalField> additionalFields) throws MessagingException, InnaccessibleFieldException {
+        Optional<MessageContent> message = fetchMessage(additionalFields, mail);
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: It's more a `messageContent`, no?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 23]
```diff
@@ -78,35 +79,38 @@ public static MailDto fromMail(Mail mail, Set<AdditionalField> additionalFields)
         }
     }
 
-    private static Optional<String> fetchTextBody(Set<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+    private static Optional<String> fetchTextBody(Set<AdditionalField> additionalFields, Optional<MessageContent> message) throws InnaccessibleFieldException {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: idem

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java [hunk 24]
```diff
@@ -78,35 +79,38 @@ public static MailDto fromMail(Mail mail, Set<AdditionalField> additionalFields)
...
-                    .map(Throwing.function(extractor::extract).sneakyThrow())
-                    .flatMap(MessageContent::getTextBody);
-        } catch (MessagingException e) {
-            throw new InnaccessibleFieldException(AdditionalField.TEXT_BODY, e);
-        }
+        return message.flatMap(MessageContent::getTextBody);
     }
 
-    private static Optional<String> fetchHtmlBody(Set<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+    private static Optional<String> fetchHtmlBody(Set<AdditionalField> additionalFields, Optional<MessageContent> message) throws InnaccessibleFieldException {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: idem

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java (hunk 19 above)
Comment (blackheaven) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: You're right, but I think this PR is already lasting way too long, can we do i tin an other issue?

Comment (rouazana) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java]: I agree, but I propose to do this refactoring once this PR is merged.

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 25]
```diff
@@ -427,9 +446,229 @@ public void retrievingAMailShouldDisplayItsInformation() throws Exception {
...
             .body("error", is(errorMessage))
-            .body("recipients", containsInAnyOrder(recipient1, recipient2));
+            .body("remoteHost", is(remoteHost))
+            .body("remoteAddr", is(remoteAddr))
+            .body("lastUpdated", is(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZ")
+                    .format(ZonedDateTime.ofInstant(lastUpdated.toInstant(), ZoneId.of("UTC")))));
+    }
+
+    @Test
+    public void retrievingAMailShouldDisplayAllAdditionalsFieldWhenRequested() throws Exception {
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: s/AdditionalsField/AdditionalFields/ (several times in this file)

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 26]
```diff
@@ -427,9 +446,229 @@ public void retrievingAMailShouldDisplayItsInformation() throws Exception {
...
+    }
+
+    @Test
+    public void retrievingAMailShouldDisplayCorrectlyEncodedHeadersInValidAdditionalsFieldWhenRequested() throws Exception {
+        when(mailRepositoryStore.get(URL_MY_REPO)).thenReturn(Optional.of(mailRepository));
+        String name = NAME_1;
+        String sender = "sender@domain";
+        String recipient1 = "recipient1@domain";
+        MimeMessage mimeMessage = MimeMessageBuilder.mimeMessageBuilder()
+                .addHeader("friend", "=?UTF-8?B?RnLDqWTDqXJpYyBNQVJUSU4=?= <fmartin@linagora.com>")
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: please anonymize this.

**Comment by chibenwa** [comment]
Just need a rebase to solve conflicts.

I'll contribute this.

**Comment by chibenwa** [comment]
Just need a rebase to solve conflicts.

I'll contribute this.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/InnaccessibleFieldException.java [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.webadmin.dto;
+
+import org.apache.james.webadmin.dto.MailDto.AdditionalField;
+
+public class InnaccessibleFieldException extends Exception {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/InnaccessibleFieldException.java]: `InaccessibleFieldException`

**Comment by chibenwa** [comment]
Force pushed to solve conflict

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 28]
```diff
@@ -721,7 +721,7 @@ public void retrievingAMailShouldFailWhenAnUnknownFieldIsRequested() throws Exce
             .statusCode(HttpStatus.BAD_REQUEST_400)
             .body("statusCode", is(400))
             .body("type", is(ErrorResponder.ErrorType.INVALID_ARGUMENT.getType()))
-            .body("message", is("The field 'nonExistingField' can't be requested in additionalFields parameter"));
+            .body("message", is("The field 'nonExistingField' can't be requested  in additionalFields parameter"));
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: why did you add this space?

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: Failing test. I thought I removed it.

**Comment by chibenwa** [comment]
test this please (build results lost)

**Comment by chibenwa** [comment]
test this please (anakin)

**Comment by chibenwa** [comment]
```
[1f1d1efefe81c32285a0af5a77a2e47503eeb40d] [ERROR] src/main/java/org/apache/james/webadmin/service/MailRepositoryStoreService.java:[41] (imports) CustomImportOrder: Wrong lexicographical order for 'org.apache.james.webadmin.dto.MailDto.AdditionalField' import. Should be before 'org.apache.james.webadmin.dto.MailKeyDTO'.
```
